compiler/rustc_hir_typeck/src/coercion.rs

```diff
@@ -1044,6 +1045,24 @@ impl<'f, 'tcx> Coerce<'f, 'tcx> {
         }
     }
 
+    fn coerce_from_pat_ty(
+        &self,
+        base: Ty<'tcx>,
+        pat: Pattern<'tcx>,
+        b: Ty<'tcx>,
+    ) -> CoerceResult<'tcx> {
+        match *b.kind() {
+            // FIXME(pattern_types): allow coercion to less restrictive patterns
+            ty::Pat(b_base, b_pat) if b_pat == pat => self.unify_and(base, b_base, identity),
+            // Coercing to anything but a pattern type only works by dismissing the pattern
+            _ => {
+                let mut coerce = self.unify_and(base, b, identity)?;
+                coerce.value.0.insert(0, Adjustment { kind: Adjust::StripPattern, target: base });
+                Ok(coerce)
+            }
+        }
+    }
```

Comment:
This coercion is incomplete, and feels unnecessarily biased towards the base type, since it leads to strange behavior when round-tripping pattern types when the coercion RHS isn't equal to the pattern itself:

```
fn foo(x: pattern_type!(i32 is 0..=1)) -> pattern_type!(i32 is 0..=1) {
    let x2 = x; // The RHS of the coercee is `?0`, which falls into the second arm of this match.
    x2 // `x2` is a `i32`.
}
```

Reply:
Leads to:

```
error[E0308]: mismatched types
 --> /home/mgx/test.rs:7:5
  |
5 | fn foo(x: pattern_type!(i32 is 0..=1)) -> pattern_type!(i32 is 0..=1) {
  |                                           --------------------------- expected `(i32) is 0..=1` because of return type
6 |     let x = x;
  |             - here the type of `x` is inferred to be `i32`
7 |     x
  |     ^ expected `(i32) is 0..=1`, found `i32`
  |
  = note: expected pattern type `(i32) is 0..=1`
                     found type `i32`
```

Reply:
I ripped it out of this PR, Some transmutes in the right places just work indeed